platform/nvidia-bluefield/byo/sonic-byo.py

```diff
@@ -0,0 +1,192 @@
+    group.add_argument('--pull', help='Docker image URL to pull')
+    group.add_argument('--file', help='Docker image gz file to load')
+    group.add_argument('--image', help='Docker image name to run')
+
+    return parser.parse_args()
+
+
+def run_cmd(cmd, verbose=False):
+    if verbose:
+        print(' '.join(cmd))
+    subprocess.run(cmd)
+
+
+def sonic_services_ctl(start, verbose):
+    services = [
+        'featured',
+        'swss',
+        'syncd',
+        'pmon',
+        'snmp',
+        'lldp',
+        'gnmi',
+        'bgp',
+        'eventd'
+    ]
+
+    systemctlops_start = ['unmask', 'enable', 'start']
+    systemctlops_stop = ['stop', 'disable', 'mask']
+
+    # Start featured last
+    if start:
+        services = services[1:] + services[:1]
+
+    print('# ' 'Starting' if start else 'Stopping', ', '.join(services))
+
+    ops = systemctlops_start if start else systemctlops_stop
+    for op in ops:
+        run_cmd(["systemctl", op] + services, verbose=verbose)
+
+
+def prepare_sonic(verbose=False):
+    print('# Preparing sonic..')
+
+    sonic_services_ctl(start=False, verbose=verbose)
+
+    print('# Loading mlx5_core driver')
+    run_cmd(["modprobe", "mlx5_core"], verbose=verbose)
+
+
+def restore_sonic(verbose=False):
+    print('# Restoring sonic..')
+    sonic_services_ctl(start=True, verbose=verbose)
+
+
+def pull_image(name):
+    try:
+        print(f'# Pulling image {name}')
+        for line in api_client.pull(name, stream=True, decode=True):
+            status = line.get('status', '')
+            progress = line.get('progress', '')
+            if progress:
+                sys.stdout.write(f'\r{status}: {progress}')
+            else:
+                sys.stdout.write(f'\r{status}')
+            sys.stdout.flush()
+        print()
+        return api_client.inspect_image(name)['Id']
+    except docker.errors.APIError as e:
+        print(f'Error pulling image: {e}')
+        return None
+
+
+def load_gz(file):
+    def chunked_file(f):
+        loaded = 0
+        total = os.path.getsize(file)
+        chunk_size = max(8192, int(total / 1000))
+        with open(file, 'rb') as f:
+            while True:
+                chunk = f.read(chunk_size)
+                if not chunk:
+                    print()
+                    break
+                loaded += len(chunk)
+                progress = loaded / total * 100
+                sys.stdout.write(f'\rLoading.. {progress:.2f}%')
+                sys.stdout.flush()
+                yield chunk
+
+    try:
+        print(f'# Loading image {file}')
+        return client.images.load(chunked_file(file))[0].id
+
+    except Exception as e:
+        print(f'Failed to load: {e}')
+        return None
+
+
+def run_container(image):
+    print(f'# Running image {image}')
+    config = {
+        'image': image,
+        'name': CONTAINER_NAME,
+        'detach': True,
+        'tty': True,
+        'privileged': True,
```

Comment:
>'privileged': True, [](http://example.com/codeflow?start=8&length=19)

running container with full privilege is dangerous, possible to limit the privileged scope?

Reply:
The idea is that the container/app takes full control over the system, so full privilege is expected